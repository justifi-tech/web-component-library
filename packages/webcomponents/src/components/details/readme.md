# justifi-details



<!-- Auto Generated Below -->


## Properties

| Property       | Attribute       | Description | Type     | Default     |
| -------------- | --------------- | ----------- | -------- | ----------- |
| `errorMessage` | `error-message` |             | `string` | `undefined` |


## Dependencies

### Used by

 - [justifi-subaccount-details](../subaccount-details)
 - [payment-details-core](../payment-details)
 - [payout-details-core](../payout-details)

### Graph
```mermaid
graph TD;
  justifi-subaccount-details --> justifi-details
  payment-details-core --> justifi-details
  payout-details-core --> justifi-details
  style justifi-details fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
