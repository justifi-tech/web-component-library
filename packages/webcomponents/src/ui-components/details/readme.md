# justifi-details



<!-- Auto Generated Below -->


## Properties

| Property       | Attribute       | Description | Type     | Default     |
| -------------- | --------------- | ----------- | -------- | ----------- |
| `errorMessage` | `error-message` |             | `string` | `undefined` |


## Dependencies

### Used by

 - [payment-details-core](../../components/payment-details)
 - [payout-details-core](../../components/payout-details)

### Graph
```mermaid
graph TD;
  payment-details-core --> justifi-details
  payout-details-core --> justifi-details
  style justifi-details fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
