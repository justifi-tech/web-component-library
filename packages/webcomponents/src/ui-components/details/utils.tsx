import { FunctionalComponent, h } from '@stencil/core';
import { heading1, heading2, text } from '../../styles/parts';

export const ErrorState = (errorMessage: string) => (
  <main
    class="p-4 d-flex justify-content-center text-center"
    style={{ fontSize: '1.2rem' }}
  >
    <p part={text}>{errorMessage}</p>
  </main>
);

export const DetailSectionTitle: FunctionalComponent<{ sectionTitle: string }> = (props) => (
  <h2
    part={heading2}
    class="mt-4 fs-5 border-bottom pb-3"
  >
    {props.sectionTitle}
  </h2>
);

export const DetailItem = ({
  title,
  value,
}: {
  title: string;
  value: string;
}) =>
  title && value ? (
    <div class="d-table-row gap-2">
      <span class="fw-bold d-table-cell px-2" part={text}>
        {title}
      </span>
      <span
        class="flex-1 d-table-cell px-2 text-wrap"
        part={text}>
        {value}
      </span>
    </div>
  ) : null;

export const EntityHeadInfoItem = ({
  title,
  value,
  classes,
}: {
  title: string;
  value: any;
  classes?: string;
}) => (
  <div class={`d-flex flex-column ${classes}`}>
    <span
      class="fw-bold border-1 border-bottom ps-2 pe-2"
      part={heading2}
    >
      {title}
    </span>
    <span class="pt-2 ps-2 pe-2" part={text}>
      {value}
    </span>
  </div>
);

export const EntityHeadInfo = ({ title, slot, badge }, children) => (
  <div slot={slot} class="p-2">
    <div class="d-flex flex-row align-items-center gap-2 mb-2">
      <h1 class="m-0" part={heading1}>
        {title}
      </h1>
      {badge}
    </div>
    <div class="d-flex flex-row align-items-top" part={text}>
      {children}
    </div>
  </div>
);

export const CodeBlock = ({ metadata }: { metadata: {} }) =>
  metadata && Object.keys(metadata).length ? (
    <div class='mt-4'>
      <pre class="p-2" aria-label="metadata content">
        <code>
          {JSON.stringify(metadata, null, 2)}
        </code>
      </pre>
    </div>
  ) : null;
