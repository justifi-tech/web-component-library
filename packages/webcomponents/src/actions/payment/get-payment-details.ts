import { Payment } from '../../api';
import { ComponentErrorSeverity } from '../../api/ComponentError';
import { getErrorCode, getErrorMessage } from '../../api/services/utils';

export const makeGetPaymentDetails =
  ({ id, authToken, service }) =>
  async ({ onSuccess, onError, final }) => {
    try {
      const response = await service.fetchPayment(id, authToken);
      if (!response.error) {
        onSuccess({ payment: new Payment(response.data) });
      } else {
        const responseError = getErrorMessage(response.error);
        const code = getErrorCode(response.error?.code);
        onError({
          error: responseError,
          code,
          severity: ComponentErrorSeverity.ERROR,
        });
      }
    } catch (error) {
      const code = getErrorCode(error?.code);
      onError({
        error: getErrorMessage(error),
        code,
        severity: ComponentErrorSeverity.ERROR,
      });
    } finally {
      final();
    }
  };
