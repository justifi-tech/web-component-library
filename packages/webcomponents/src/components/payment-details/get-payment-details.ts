import { Payment } from '../../api';
import { getErrorMessage } from '../../api/services/utils';

export const makeGetPaymentDetails =
  ({ id, authToken, service }) =>
  async ({ onSuccess, onError }) => {
    try {
      const response = await service.fetchPayment(id, authToken);
      if (!response.error) {
        onSuccess({ payment: new Payment(response.data) });
      } else {
        const responseError = getErrorMessage(response.error);
        onError(`Error trying to fetch data : ${responseError}`);
      }
    } catch (error) {
      onError(`Error trying to fetch data : ${error}`);
    }
  };
