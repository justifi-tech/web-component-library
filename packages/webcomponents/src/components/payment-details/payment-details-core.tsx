import { Component, h, Prop, State, Watch, Event, EventEmitter } from '@stencil/core';
import { Payment } from '../../api';
import { formatCurrency, formatDate, formatTime, snakeCaseToHumanReadable } from '../../utils/utils';
import { CodeBlock, DetailItem, DetailSectionTitle, EntityHeadInfo, EntityHeadInfoItem, ErrorState, LoadingState } from '../../ui-components/details/utils';
import { ComponentError } from '../../api/ComponentError';
import { StyledHost } from '../../ui-components';
import { MapPaymentStatusToBadge } from '../payments-list/payments-status';

@Component({
  tag: 'payment-details-core',
})

export class PaymentDetailsCore {
  @Prop() getPaymentDetails: Function;

  @State() payment: Payment;
  @State() loading: boolean = true;
  @State() errorMessage: string;

  @Event({ eventName: 'error-event' }) errorEvent: EventEmitter<ComponentError>

  componentWillLoad() {
    if (this.getPaymentDetails) {
      this.fetchData();
    }
  }

  @Watch('getPaymentDetails')
  updateOnPropChange() {
    this.fetchData();
  }

  fetchData(): void {
    this.loading = true;

    this.getPaymentDetails({
      onSuccess: ({ payment }) => {
        this.payment = payment;
        this.loading = false;
        this.errorMessage = null;
      },
      onError: ({ error, code, severity }) => {
        this.errorMessage = error;
        this.errorEvent.emit({
          errorCode: code,
          message: error,
          severity,
        })
        this.loading = false;
      },
    });
  }

  render() {
    return (
      <StyledHost>
        {this.loading && LoadingState()}
        {!this.loading && this.errorMessage && ErrorState(this.errorMessage)}
        {!this.loading && !this.errorMessage &&
          this.payment && (
            <justifi-details error-message={this.errorMessage}>
              <EntityHeadInfo
                slot="head-info"
                badge={<span slot='badge' part="detail-head-badge" innerHTML={MapPaymentStatusToBadge(this.payment.status)} />}
                title={`${formatCurrency(this.payment.amount)} ${this.payment?.currency.toUpperCase()}`}
              >
                <EntityHeadInfoItem
                  classes="border-1 border-end"
                  title="Updated At"
                  value={`${formatDate(this.payment.updated_at)} ${formatTime(this.payment.updated_at)}`}
                />
                <EntityHeadInfoItem
                  classes="border-1 border-end"
                  title="Created At"
                  value={`${formatDate(this.payment.created_at)} ${formatTime(this.payment.created_at)}`}
                />
                <EntityHeadInfoItem title="ID" value={this.payment.id} />
              </EntityHeadInfo>
              <div slot='detail-sections' part="detail-section">
                <DetailSectionTitle sectionTitle="Details" />
                <div class="d-table gap-2 w-100">
                  <DetailItem title="Amount" value={formatCurrency(this.payment.amount)} />
                  <DetailItem title="Fees" value={formatCurrency(this.payment.fee_amount)} />
                  <DetailItem title="Refunded" value={formatCurrency(this.payment.amount_refunded)} />
                  <DetailItem title="Net" value={formatCurrency(this.payment.balance)} />
                  <DetailItem title="Status" value={MapPaymentStatusToBadge(this.payment.status)} />
                  <DetailItem title="Payment ID" value={this.payment.id} />
                  <DetailItem title="Processing Fees" value={formatCurrency(this.payment.fee_amount)} />
                  <DetailItem title="Statement Descriptor" value={this.payment.statement_descriptor} />
                  <DetailItem title="Description" value={this.payment.description} />
                </div>
                {this.payment.payment_method.card && [
                  <DetailSectionTitle sectionTitle="Payment Method" />,
                  <div class="d-table gap-2 w-100">
                    <DetailItem title="ID" value={this.payment.payment_method.card.id} />
                    <DetailItem title="Payment Type" value="Card" />
                    <DetailItem title="Last 4 Numbers" value={this.payment.payment_method.lastFourDigits} />
                    <DetailItem title="Brand" value={snakeCaseToHumanReadable(this.payment.payment_method.card.brand)} />
                    <DetailItem title="Cardholder" value={this.payment.payment_method.payersName} />
                  </div>
                ]}
                {this.payment.payment_method.bank_account && [
                  <DetailSectionTitle sectionTitle="Payment Method" />,
                  <div class="d-table gap-2 w-100">
                    <DetailItem title="ID" value={this.payment.payment_method.bank_account.id} />
                    <DetailItem title="Last 4 Numbers" value={this.payment.payment_method.lastFourDigits} />
                    <DetailItem title="Bank Name" value={this.payment.payment_method.bank_account.brand} />
                    <DetailItem title="Account Owner" value={this.payment.payment_method.payersName} />
                  </div>
                ]}
                <DetailSectionTitle sectionTitle='Metadata' />
                <div class="d-table gap-2 w-100">
                  <CodeBlock metadata={this.payment.metadata} />
                </div>
              </div>
            </justifi-details>
          )}
      </StyledHost>
    );
  }
}
