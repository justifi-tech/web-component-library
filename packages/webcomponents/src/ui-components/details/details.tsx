import { Component, h, Prop } from '@stencil/core';
import { ErrorState } from './utils';

@Component({
  tag: 'justifi-details',
})

export class Details {
  @Prop() errorMessage: string;

  render() {
    return (
      <div>
        {
          this.errorMessage ?
            ErrorState(this.errorMessage)
            :
            <main class="p-2">
              <slot name="head-info" />
              <slot name='detail-sections' />
            </main>
        }
      </div>
    )
  }
}
