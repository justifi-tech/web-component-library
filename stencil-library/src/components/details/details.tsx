import { Component, Host, h, Prop } from '@stencil/core';
import { ErrorState } from './utils';

@Component({
  tag: 'justifi-details',
  styleUrl: './details.scss',
  shadow: true,
})

export class Details {
  @Prop() errorMessage: string;

  render() {
    return (
      <Host
        exportParts='detail-loading-spinner,detail-loading-state,detail-empty-state,
        detail-head,detail-title,detail-head-info,detail-head-info-item,
        detail-head-info-item-title,detail-head-info-item-data,detail-metadata,detail-metadata-title,
        detail-section,detail-section-title,detail-section-item-title,detail-section-item-data'
      >
        {
          this.errorMessage ?
            ErrorState(this.errorMessage)
            :
            <main class="p-2">
              <slot name="head-info" />
              <slot name='detail-sections' />
            </main>
        }
      </Host>
    )
  }
}
