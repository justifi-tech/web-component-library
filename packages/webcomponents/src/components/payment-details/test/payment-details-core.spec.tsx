import { newSpecPage } from '@stencil/core/testing';
import mockPaymentDetailsResponse from '../../../api/mockData/mockPaymentDetailSuccess.json';
import { PaymentDetailsCore } from '../payment-details-core';
import { Details } from '../../details/details';
import { makeGetPaymentDetails } from '../get-payment-details';

describe('payment-details-core', () => {
  const components = [PaymentDetailsCore, Details];
  it('it renders properly with fetched data', async () => {
    const mockPaymentService = {
      fetchPayment: jest.fn().mockResolvedValue(mockPaymentDetailsResponse),
    };

    const page = await newSpecPage({
      components,
      html: `<payment-details-core></payment-details-core>`,
    });

    page.rootInstance.componentWillLoad = () => { };

    page.rootInstance.getPaymentDetails = makeGetPaymentDetails({
      id: '123',
      authToken: '123',
      service: mockPaymentService
    });

    page.rootInstance.fetchData();
    await page.waitForChanges();

    expect(page.rootInstance.payment).toEqual(expect.objectContaining({ id: mockPaymentDetailsResponse.data.id }));
    expect(mockPaymentService.fetchPayment).toHaveBeenCalled();
    // expect(page.root).toMatchSnapshot();
  });

  it('displays an error state on failed data fetch', async () => {
    const mockService = {
      fetchPayment: jest.fn().mockRejectedValue(new Error('Fetch error'))
    };

    const getPaymentDetails = makeGetPaymentDetails({
      id: 'some-id',
      authToken: 'some-auth-token',
      service: mockService
    });

    const page = await newSpecPage({
      components,
      html: '<payment-details-core></payment-details-core>',
    });

    page.rootInstance.componentWillLoad = () => { };

    page.rootInstance.getPaymentDetails = getPaymentDetails;
    page.rootInstance.fetchData();
    await page.waitForChanges();

    // expect(page.root).toMatchSnapshot();
  });
});
