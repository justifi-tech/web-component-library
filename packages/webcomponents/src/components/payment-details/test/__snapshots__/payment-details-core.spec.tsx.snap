// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`payment-details-core displays an error state on failed data fetch 1`] = `
<payment-details-core class="body" style="display: block;">
  <style></style>
  <main class="d-flex justify-content-center p-4 text-center" style="font-size: 1.2rem;">
    <p part="text color font-family">
      Fetch error
    </p>
  </main>
</payment-details-core>
`;

exports[`payment-details-core it renders properly with fetched data 1`] = `
<payment-details-core class="body" style="display: block;">
  <style></style>
  <justifi-details>
    <!---->
    <main class="p-2">
      <div class="p-2" slot="head-info">
        <div class="align-items-center d-flex flex-row gap-2 mb-2">
          <h1 class="m-0" part="heading-1 heading text color font-family">
            $234.56 USD
          </h1>
          <span class="badge bg-success" part="badge font-family badge-success" title="This payment was successfully captured">
            Succeeded
          </span>
        </div>
        <div class="align-items-top d-flex flex-row" part="text color font-family">
          <div class="border-1 border-end d-flex flex-column">
            <span class="border-1 border-bottom fw-bold pe-2 ps-2" part="heading-2 heading text color font-family">
              Updated At
            </span>
            <span class="pe-2 ps-2 pt-2" part="text color font-family">
              Dec 27, 2023 9:53pm
            </span>
          </div>
          <div class="border-1 border-end d-flex flex-column">
            <span class="border-1 border-bottom fw-bold pe-2 ps-2" part="heading-2 heading text color font-family">
              Created At
            </span>
            <span class="pe-2 ps-2 pt-2" part="text color font-family">
              Dec 27, 2023 9:52pm
            </span>
          </div>
          <div class="d-flex flex-column undefined">
            <span class="border-1 border-bottom fw-bold pe-2 ps-2" part="heading-2 heading text color font-family">
              ID
            </span>
            <span class="pe-2 ps-2 pt-2" part="text color font-family">
              py_1NNeEnf4FbelxDCQN2RHcE
            </span>
          </div>
        </div>
      </div>
      <div slot="detail-sections">
        <h2 class="border-bottom fs-5 mt-4 pb-3" part="heading-2 heading text color font-family">
          Details
        </h2>
        <div class="d-table gap-2 w-100">
          <div class="d-table-row gap-2">
            <span class="d-table-cell fw-bold px-2" part="text color font-family">
              Amount
            </span>
            <span class="d-table-cell flex-1 px-2 text-wrap" part="text color font-family">
              $234.56 USD
            </span>
          </div>
          <div class="d-table-row gap-2">
            <span class="d-table-cell fw-bold px-2" part="text color font-family">
              Fees
            </span>
            <span class="d-table-cell flex-1 px-2 text-wrap" part="text color font-family">
              $2.00 USD
            </span>
          </div>
          <div class="d-table-row gap-2">
            <span class="d-table-cell fw-bold px-2" part="text color font-family">
              Refunded
            </span>
            <span class="d-table-cell flex-1 px-2 text-wrap" part="text color font-family">
              $0.00 USD
            </span>
          </div>
          <div class="d-table-row gap-2">
            <span class="d-table-cell fw-bold px-2" part="text color font-family">
              Net
            </span>
            <span class="d-table-cell flex-1 px-2 text-wrap" part="text color font-family">
              $232.56 USD
            </span>
          </div>
          <div class="d-table-row gap-2">
            <span class="d-table-cell fw-bold px-2" part="text color font-family">
              Status
            </span>
            <span class="d-table-cell flex-1 px-2 text-wrap" part="text color font-family">
              <span class="badge bg-success" part="badge font-family badge-success" title="This payment was successfully captured">
                Succeeded
              </span>
            </span>
          </div>
          <div class="d-table-row gap-2">
            <span class="d-table-cell fw-bold px-2" part="text color font-family">
              Payment ID
            </span>
            <span class="d-table-cell flex-1 px-2 text-wrap" part="text color font-family">
              py_1NNeEnf4FbelxDCQN2RHcE
            </span>
          </div>
          <div class="d-table-row gap-2">
            <span class="d-table-cell fw-bold px-2" part="text color font-family">
              Processing Fees
            </span>
            <span class="d-table-cell flex-1 px-2 text-wrap" part="text color font-family">
              $2.00 USD
            </span>
          </div>
          <div class="d-table-row gap-2">
            <span class="d-table-cell fw-bold px-2" part="text color font-family">
              Description
            </span>
            <span class="d-table-cell flex-1 px-2 text-wrap" part="text color font-family">
              payment
            </span>
          </div>
        </div>
        <h2 class="border-bottom fs-5 mt-4 pb-3" part="heading-2 heading text color font-family">
          Payment Method
        </h2>
        <div class="d-table gap-2 w-100">
          <div class="d-table-row gap-2">
            <span class="d-table-cell fw-bold px-2" part="text color font-family">
              ID
            </span>
            <span class="d-table-cell flex-1 px-2 text-wrap" part="text color font-family">
              pm_uICmv8696AatL4gRwa0Wq
            </span>
          </div>
          <div class="d-table-row gap-2">
            <span class="d-table-cell fw-bold px-2" part="text color font-family">
              Last 4 Numbers
            </span>
            <span class="d-table-cell flex-1 px-2 text-wrap" part="text color font-family">
              **** 6789
            </span>
          </div>
          <div class="d-table-row gap-2">
            <span class="d-table-cell fw-bold px-2" part="text color font-family">
              Bank Name
            </span>
            <span class="d-table-cell flex-1 px-2 text-wrap" part="text color font-family">
              Altra
            </span>
          </div>
          <div class="d-table-row gap-2">
            <span class="d-table-cell fw-bold px-2" part="text color font-family">
              Account Owner
            </span>
            <span class="d-table-cell flex-1 px-2 text-wrap" part="text color font-family">
              Julia Garnett
            </span>
          </div>
        </div>
      </div>
    </main>
  </justifi-details>
</payment-details-core>
`;
